File: e2e/tests/consent.spec.ts
import { test, expect } from '@playwright/test';
import { setup, startAuthorization } from './helpers';

// force_consent=1 disables the development-mode skip_authorization, so the
// real Doorkeeper consent screen renders. Each test runs in a fresh browser
// context, so the session-stored current_user never leaks between tests.
test.describe('Consent screen', () => {
  test('approving issues a code that can be exchanged', async ({ page }) => {
    const s = await setup(page, 'consent-ok');
    await startAuthorization(page, s, { forceConsent: true });

    await expect(page.locator('body')).toContainText('Authorization required');
    await expect(page.locator('body')).toContainText(s.app.name);

    await page.getByRole('button', { name: 'Authorize' }).click();

    await expect(page.locator('#authorize_result')).toContainText('"code"');
    await page.getByRole('button', { name: 'Exchange' }).click();
    await expect(page.locator('#exchange_result')).toContainText('"access_token"');
  });

  test('denying redirects back with access_denied', async ({ page }) => {
    const s = await setup(page, 'consent-ng');
    await startAuthorization(page, s, { forceConsent: true });

    await expect(page.locator('body')).toContainText('Authorization required');
    await page.getByRole('button', { name: 'Deny' }).click();

    await expect(page.locator('#authorize_result')).toContainText('access_denied');
    await expect(page.locator('#authorize_result')).not.toContainText('"code"');
  });
});

File: spec/e2e/consent_spec.rb
# frozen_string_literal: true

require_relative "e2e_helper"

# force_consent=1 turns off the development-only skip_authorization, so
# Doorkeeper's real consent screen renders. Capybara resets the session between
# examples, so the remembered current_user never leaks from one to the next.
RSpec.describe "Consent screen" do
  it "issues a code that can be exchanged when the user approves" do
    client = dashboard_client("consent-ok")
    start_authorization(client, force_consent: true)

    expect(page).to have_text("Authorization required")
    expect(page).to have_text(client.name)

    click_button "Authorize"

    expect(page).to have_css("#authorize_result", text: '"code"')
    click_button "Exchange"
    expect(page).to have_css("#exchange_result", text: '"access_token"')
  end

  # The consent form is the one place a nonce can silently go missing: the
  # browser posts back only what the view carried as hidden fields, so a view
  # without a nonce field issues an ID token without one, exactly the trap
  # behind #154. Drive it through a real browser rather than trusting the
  # rendered markup alone.
  it "carries the nonce through the consent screen into the ID token" do
    client = dashboard_client("consent-nonce")
    nonce = start_authorization(client, nonce: true, force_consent: true)

    expect(page).to have_text("Authorization required")
    expect(page).to have_css("input[name='nonce'][value='#{nonce}']", visible: :all)
    click_button "Authorize"

    expect(page).to have_css("#authorize_result", text: '"code"')
    click_button "Exchange"
    expect(page).to have_css("#exchange_result", text: '"id_token"')
    expect(id_token_payload["nonce"]).to eq(nonce)
  end

  it "redirects back with access_denied when the user denies" do
    client = dashboard_client("consent-ng")
    start_authorization(client, force_consent: true)

    expect(page).to have_text("Authorization required")

    click_button "Deny"

    expect(page).to have_css("#authorize_result", text: "access_denied")
    expect(page).to have_no_css("#authorize_result", text: '"code"')
  end
end
